head.go

```diff
@@ -752,22 +762,89 @@ func (h *Head) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 		}
 		// Delete only until the current values and not beyond.
 		t0, t1 = clampInterval(mint, maxt, t0, t1)
-		stones = append(stones, Stone{p.At(), Intervals{{t0, t1}}})
+		if err := h.delete(p.At(), Intervals{{t0, t1}}); err != nil {
+			return errors.Wrap(err, "delete samples")
+		}
+		dirty = true
 	}
-
 	if p.Err() != nil {
 		return p.Err()
 	}
-	var enc RecordEncoder
+	if dirty {
+		h.gc()
+	}
+
+	return nil
+}
+
+// delete re-writes the chunks which overlaps with deleted ranges
+// and removes the samples in the deleted ranges. Chunks is deleted
+// if no samples are left at the end.
+func (h *Head) delete(ref uint64, dranges Intervals) error {
+	if len(dranges) == 0 {
+		return nil
+	}
+
+	stripeSeriesIndex := ref % stripeSize
+
+	h.series.locks[stripeSeriesIndex].Lock()
+	defer h.series.locks[stripeSeriesIndex].Unlock()
+	ms, ok := h.series.series[stripeSeriesIndex][ref]
+	if !ok {
+		return fmt.Errorf("Mem series not found for reference id %d", ref)
+	}
+	ms.Lock()
+	defer ms.Unlock()
 
-	if h.wal != nil {
-		if err := h.wal.Log(enc.Tombstones(stones, nil)); err != nil {
+	// Deleting samples for intervals in tombstones.
+	var chnksToDel []int
+	for i, chk := range ms.chunks {
+		if !chk.OverlapsClosedInterval(dranges[0].Mint, dranges[len(dranges)-1].Maxt) {
+			continue
+		}
+
+		// Create new chunk by removing deleted samples.
+		newChunk := chunkenc.NewXORChunk()
+		app, err := newChunk.Appender()
+		if err != nil {
 			return err
 		}
+
+		it := &deletedIterator{it: chk.chunk.Iterator(), intervals: dranges}
+		for it.Next() {
+			ts, v := it.At()
+			app.Append(ts, v)
+		}
+
+		if newChunk.NumSamples() > 0 {
+			ms.chunks[i].chunk = newChunk
+		} else { // Empty chunk, not required to store.
+			chnksToDel = append(chnksToDel, i)
+		}
 	}
-	for _, s := range stones {
-		h.tombstones.addInterval(s.ref, s.intervals[0])
+	// Deleting empty chunks.
+	for i, idx := range chnksToDel {
+		ms.chunks = append(ms.chunks[:idx-i], ms.chunks[idx-i+1:]...)
+	}
+
+	// Getting the last 4 samples in the last chunk
+	// to update the sampleBuf.
```

Comment:
I couldn't fully understand the logic behind `sampleBuf`(we need some better comments somewhere about this), but shouldn't this update happen only when we have some modification in the last chunk?
Track if the delete happened in the last chunk and only than update the buffer?

Ideally this should be in a separate func `updateSampleBuf`  as well.

Reply:
I was just checking last 4 samples of the last chunk and updating it, that's what `sampleBuf` stores.

But now after the latest change we wont require this.